### PR TITLE
feat: toast component migration

### PR DESCRIPTION
**Pull request by jvorcak**
This PR migrates the message and toast component to use Chakra's Toast implementation.

WIP: I'm in a middle of testing, but I think this PR can be already looked at.

- Chakra's Toast is implemented using a hook, so a functional component is required. 
  1. On a couple of places, I've migrated the component to a FC, this needs to be tested
  2. At the same time, I wanted to minimise the number of changes, so on some places I've left React as a component class and left a comment about a future migration. In these cases, I use Chakra's standalone toasts.
  3. A couple of messages are not migrated, because of the problem with Modals. Chakra's toast and Ant design modals doesn't play well together, because of ChakraProvider that kills the toasts on its removal. This will be addressed together with a modal migration.

Screenshot 2023-09-19 at 7 38 43

There are a couple of "special" places in the platform worth checking out:

-  Using Toast as a loader indicator
 Video:  https://www.loom.com/share/9749918a941b497688c8a0b56ea72258
 Code: https://github.com/redpanda-data/console/pull/846/files#diff-b8d5e9131eed247da5d345f825ecaccfed927cb04939825ae7a8335374bc7819
- Error screen
 Screen: 
Screenshot 2023-09-19 at 8 00 43

- Updating existing Toast
   https://www.loom.com/share/4127a1ac5c8144a6ad8462a35607af4a?from_recorder=1&focus_title=1

### CODE REVIEW

File: frontend/src/components/misc/KowlJsonView.tsx [hunk 1]
```diff
@@ -133,15 +133,15 @@ export const KowlJsonView = observer((props: ReactJsonViewProps) => {
                         collapsed={settings.collapsed}
                         onSelect={e => {
                             if (ctrlDown) {
-                                if (navigator?.clipboard) {
-                                    navigator.clipboard.writeText(String(e.value));
-                                    message.success(
-                                        <span>
-                                            Copied value of <span className="codeBox">{e.name}</span>
+                                navigator.clipboard.writeText(String(e.value)).then(() => {
+                                    toast({
```
Comment (rikimaru0345) [inline frontend/src/components/misc/KowlJsonView.tsx:137]: should have `.catch(...` or checking if clipboard is available; there are cases where its not

File: frontend/package.json [hunk 2]
```diff
@@ -28,7 +28,7 @@
         "@heroicons/react": "^1.0.6",
         "@monaco-editor/react": "^4.5",
         "@primer/octicons-react": "^17",
-        "@redpanda-data/ui": "^3.17.2",
+        "@redpanda-data/ui": "^3.21.0",
```
Comment (malinskibeniamin) [inline frontend/package.json:31]: For the latest UI release (3.24.0 at the time of writing this), there have been changes to the Text component as well as the color palette, please make sure that the look & feel is still good.

File: frontend/src/components/misc/HideStatisticsBarButton.tsx [hunk 3]
```diff
@@ -9,26 +9,25 @@
...
-                    </div>
-                </div>
-            </Tooltip>
-        );
-    }
+export const HideStatisticsBarButton: FC<{ onClick: Function }> = ({onClick}) => {
+    const toast = useToast()
+    return (
+        <Tooltip label={<Text whiteSpace="nowrap">Hide statistics bar</Text>} placement="right" hasArrow>
+            <div className="hideStatsBarButton" onClick={() => {
```
Comment (malinskibeniamin) [inline frontend/src/components/misc/HideStatisticsBarButton.tsx:20]: [suggestion] I would prefer if we use a Box component with the classname styles applied accordingly, if it is possible, since this is relevant for the toast migration in this example

File: frontend/src/components/misc/UserPreferences.tsx [hunk 4]
```diff
@@ -156,86 +158,101 @@ class JsonViewerTab extends Component {
...
+                    console.error('unable to copy settings to clipboard', {error: e});
+                })
+            }}>
+                Export User Preferences
+            </Button>
+        </Label>
+
+        <Label text="Reset">
+            <>
+                <div>
```
Comment (malinskibeniamin) [inline frontend/src/components/misc/UserPreferences.tsx:237]: We may not need this div

File: frontend/src/components/misc/UserPreferences.tsx [hunk 5]
```diff
@@ -156,86 +158,101 @@ class JsonViewerTab extends Component {
...
+                           value={$state.resetConfirm}
+                           onChange={str => $state.resetConfirm = str.target.value} />
+                    <Button onClick={() => {
+                        clearSettings();
+                        toast({
+                            status: 'success',
+                            description: 'All settings have been reset to their defaults'
+                        });
+                        $state.resetConfirm = '';
+                    }} colorScheme="red" isDisabled={$state.resetConfirm !== 'reset'}>Reset</Button>
```
Comment (malinskibeniamin) [inline frontend/src/components/misc/UserPreferences.tsx:249]: [question] I think the `colorScheme` should be `brand`?

File: frontend/src/components/misc/UserPreferences.tsx [hunk 6]
```diff
@@ -156,86 +158,101 @@ class JsonViewerTab extends Component {
...
+                </div>
+                <span className="smallText">Clear all your user settings, resetting them to the default values</span>
```
Comment (malinskibeniamin) [inline frontend/src/components/misc/UserPreferences.tsx:251]: [suggestion] [future] After we start using 3.24.0, maybe we can start using the Text component more that will eventually have variants, provided by @frankieshakes

File: frontend/src/components/pages/acls/Acl.List.tsx [hunk 7]
```diff
@@ -30,7 +30,16 @@ import PageContent from '../../misc/PageContent';
 import createAutoModal from '../../../utils/createAutoModal';
 import { CreateServiceAccountEditor, generatePassword } from './CreateServiceAccountEditor';
 import { Features } from '../../../state/supportedFeatures';
-import { Alert, AlertIcon, Badge, Button, Icon, SearchField, Tooltip } from '@redpanda-data/ui';
+import { Alert, AlertIcon, Badge, Button, createStandaloneToast, Icon, redpandaToastOptions, SearchField, Tooltip } from '@redpanda-data/ui';
+
+// TODO - once AclList is migrated to FC, we could should move this code to use useToast()
+const { ToastContainer, toast } = createStandaloneToast({
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/acls/Acl.List.tsx:36]: 👍

File: frontend/src/components/pages/acls/Acl.List.tsx [hunk 8]
```diff
@@ -86,7 +95,10 @@ class AclList extends PageComponent {
                                 operation: 'Any',
                                 permissionType: 'Any',
                             });
-                            message.success(<>Deleted ACLs for <Code>{record.principalName}</Code></>);
+                            toast({
+                                status: 'success',
+                                description: <>Deleted ACLs for <Code>{record.principalName}</Code></>
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/acls/Acl.List.tsx:100]: Let's use Text as="span"

File: frontend/src/components/misc/HideStatisticsBarButton.tsx [hunk 9]
```diff
@@ -9,26 +9,25 @@
...
-                <div className="hideStatsBarButton" onClick={this.handleClick}>
-                    <div style={{ display: 'flex', width: '100%' }}>
-                        <EyeClosedIcon size="medium" />
-                    </div>
-                </div>
-            </Tooltip>
-        );
-    }
+export const HideStatisticsBarButton: FC<{ onClick: Function }> = ({onClick}) => {
+    const toast = useToast()
```
[the remark this answers is not in the document]
Comment (malinskibeniamin) [reply frontend/src/components/misc/HideStatisticsBarButton.tsx:17]: Resolved, discussed during meeting with @frankieshakes

File: frontend/src/components/misc/UserPreferences.tsx (hunk 5 above)
Comment (malinskibeniamin) [reply frontend/src/components/misc/UserPreferences.tsx:249]: Nevermind, for now let's leave it until Button variant updates.

File: frontend/src/utils/tsxUtils.tsx [hunk 10]
```diff
@@ -385,45 +394,53 @@ export class StatusIndicator extends Component<StatusIndicatorProps> {
...
-                            isIndeterminate={this.props.statusText === 'Connecting'}
-                            colorScheme="brand"
-                        />
-                    </Box>
+            <Box mb="0.2em">
+                <Box minW={300}>
+                    <Progress
+                        value={this.props.fillFactor * 100}
+                        isIndeterminate={this.props.statusText === 'Connecting'}
+                        colorScheme="blue"
```
Comment (malinskibeniamin) [inline frontend/src/utils/tsxUtils.tsx:408]: [question] Here color scheme is blue, but before it was brand, is that intentional?

File: frontend/src/components/pages/topics/Tab.Messages/index.tsx [hunk 11]
```diff
@@ -65,6 +65,51 @@ interface TopicMessageViewProps {
...
+                navigator.clipboard.writeText(getStringValue(record)).then(() => {
+                    toast({
+                        status: 'success',
+                        description: 'Key copied to clipboard'
+                    })
+                })
+            }}>
+                Copy Key
+            </Menu.Item>
+            <Menu.Item key="2" disabled={record.value.isPayloadNull} onClick={() => {
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/topics/Tab.Messages/index.tsx:85]: Why not key 1, 2, 3, 4?

File: frontend/src/components/pages/schemas/Schema.Details.tsx [hunk 12]
```diff
@@ -166,121 +171,125 @@ class SchemaDetailsView extends PageComponent<{ subjectName: string }> {
...
+                                        {title: 'Default', dataIndex: 'default', className: 'whiteSpaceDefault'},
+                                        {title: 'Documentation', dataIndex: 'doc', className: 'whiteSpaceDefault'},
+                                    ]}
+                                    rowKey="name"
+                                    dataSource={fields}
+                                    pagination={false}
+                                    style={{
+                                        maxWidth: '100%',
+                                        marginTop: '1.5rem',
+                                        marginBottom: '1.5rem',
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/schemas/Schema.Details.tsx:284]: If you use Box, you can just do `my="1.5rem"`

File: frontend/src/components/pages/schemas/Schema.Details.tsx [hunk 13]
```diff
@@ -166,121 +171,125 @@ class SchemaDetailsView extends PageComponent<{ subjectName: string }> {
...
+                                         onChange={s => uiSettings.schemaDetails.viewMode = s}
                             />
-                        }
 
-                    </div>
-                </Section>
-            </PageContent>
+                            <NoClipboardPopover placement="top">
+                                <div>
+                                    {' '}
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/schemas/Schema.Details.tsx:235]: [suggestion]
```suggestion
```

File: frontend/src/components/pages/reassign-partitions/components/ActiveReassignments.tsx [hunk 14]
```diff
@@ -240,18 +199,71 @@ export class ThrottleDialog extends Component<{ visible: boolean, lastKnownMinTh
...
+        <div style={{ display: 'flex', flexDirection: 'column', gap: '1em', }}>
+            <div style={{ margin: '0 1em' }}>
+                <p style={{ margin: 0 }}>Using throttling you can limit the network traffic for reassignments.</p>
+                <ul style={{ marginTop: '0.5em', padding: '0 1.5em' }}>
+                    <li>Throttling applies to all replication traffic, not just to active reassignments.</li>
+                    <li>Once the reassignment completes you'll have to remove the throttling configuration. <br />
+                        Console will show a warning below the "Current Reassignments" table when there are throttled topics that are no longer being reassigned.
+                    </li>
+                </ul>
+            </div>
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/reassign-partitions/components/ActiveReassignments.tsx:259]: For cases like these, if it's not too much work, using `UnorderedList` and `ListItem` should just work out of the box, so consider it (of course can be a separate PR to avoid scope creep)

File: frontend/src/components/pages/reassign-partitions/components/ActiveReassignments.tsx [hunk 15]
```diff
@@ -155,81 +155,40 @@ export class ActiveReassignments extends Component<{ throttledTopics: string[],
...
+    const toastRef = useRef<ToastId>()
 
-        </Modal>
-    }
+    const throttleValue = $state.newThrottleValue ?? 0;
+    const noChange = ($state.newThrottleValue === lastKnownMinThrottle)
+        || ($state.newThrottleValue == null);
 
-    async applyBandwidthThrottle() {
+    // console.log('nochange:', { noChange, newVal: this.newThrottleValue, lastKnown: this.props.lastKnownMinThrottle })
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/reassign-partitions/components/ActiveReassignments.tsx:172]: Remove `console.log`
```suggestion
```

File: frontend/src/utils/tsxUtils.tsx (hunk 10 above)
Comment (jvorcak) [reply frontend/src/utils/tsxUtils.tsx:408]: @malinskibeniamin for now we only have the following variants for a Toast.
Screenshot 2023-09-25 at 12 54 00

So I had to change the progress to match the (blue) neutral colour. We can later come up with some neutral (gray) variant in the future? But I think it's worth revisiting once we have a styleguide? WDYT? 

Screenshot 2023-09-25 at 12 55 57

File: frontend/src/components/pages/topics/Tab.Messages/index.tsx (hunk 11 above)
Comment (jvorcak) [reply frontend/src/components/pages/topics/Tab.Messages/index.tsx:85]: @malinskibeniamin this was there from before, I can surely refactor all of this, the question is to what level should I dive when refactoring :) Right now I do only what is necessary to migrate away from Antd and inline styles, but if you want I can go deeper by default.

File: frontend/src/components/pages/schemas/Schema.Details.tsx (hunk 12 above)
Comment (jvorcak) [reply frontend/src/components/pages/schemas/Schema.Details.tsx:284]: @malinskibeniamin I was thinking I'd migrate these styles when migrating the table component?

File: frontend/src/components/pages/reassign-partitions/components/ActiveReassignments.tsx (hunk 14 above)
Comment (jvorcak) [reply frontend/src/components/pages/reassign-partitions/components/ActiveReassignments.tsx:258]: I think I've already addressed this in the #847, but I'll add a note there to double check

File: frontend/src/components/misc/KowlJsonView.tsx (hunk 1 above)
Comment (jvorcak) [reply frontend/src/components/misc/KowlJsonView.tsx:138]: Fixed

File: frontend/src/components/pages/topics/Tab.Messages/index.tsx (hunk 11 above)
Comment (malinskibeniamin) [reply frontend/src/components/pages/topics/Tab.Messages/index.tsx]: No worries, just asking - please feel free to leave it for now.

File: frontend/src/components/pages/schemas/Schema.Details.tsx (hunk 12 above)
Comment (malinskibeniamin) [reply frontend/src/components/pages/schemas/Schema.Details.tsx:283]: Makes sense, that sounds good to me, @rikimaru0345 any style guide to follow here?

File: frontend/src/components/misc/ErrorBoundary.tsx [hunk 16]
```diff
@@ -219,6 +217,29 @@ function getStringFromInfo(info: InfoItem) {
...
+    const toast = useToast()
+
+    return (
+        <Button
+            variant="ghost"
+            size="large"
+            disabled={disabled}
+            isLoading={isLoading}
+            onClick={() => {
+                navigator.clipboard.writeText(message).then(() => {
```
Comment (malinskibeniamin) [inline frontend/src/components/misc/ErrorBoundary.tsx:230]: [question] Will this work on non-Chrome browsers? https://caniuse.com/?search=writeText

File: frontend/src/utils/tsxUtils.tsx [hunk 17]
```diff
@@ -545,3 +562,12 @@ export function IconButton(p: { onClick?: React.MouseEventHandler<HTMLElement>;
         </Tooltip>
     );
 }
+
+
+export const navigatorClipboardErrorHandler = (e: DOMException) => {
```
Comment (malinskibeniamin) [inline frontend/src/utils/tsxUtils.tsx:567]: We may want to check which browser support this https://caniuse.com/?search=writeText

File: frontend/src/utils/tsxUtils.tsx (hunk 10 above)
Comment (malinskibeniamin) [reply frontend/src/utils/tsxUtils.tsx:408]: I will leave this one for @rikimaru0345 and @frankieshakes to decide

File: frontend/src/components/pages/topics/Topic.List.tsx [hunk 18]
```diff
@@ -336,9 +337,12 @@ function ConfirmDeletionModal({ topicToDelete, onFinish, onCancel }: { topicToDe
     const finish = () => {
         onFinish();
         cleanup();
-        notification['success']({
-            message: <>Topic <Code>{topicToDelete?.topicName}</Code> deleted successfully</>,
-        });
+
+        toast({
+            title: 'Topic Deleted',
+            description: <>Topic <Code>{topicToDelete?.topicName}</Code> deleted successfully</>,
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/topics/Topic.List.tsx:343]: [suggestion] Use `Text` component instead of `Fragment`

File: frontend/src/components/pages/schemas/Schema.Details.tsx [hunk 19]
```diff
@@ -166,121 +171,124 @@ class SchemaDetailsView extends PageComponent<{ subjectName: string }> {
...
+                                         onChange={s => uiSettings.schemaDetails.viewMode = s}
                             />
-                        }
 
-                    </div>
-                </Section>
-            </PageContent>
+                            <NoClipboardPopover placement="top">
+                                <div>
+                                    {/* the additional div is necessary because popovers do not trigger on disabled elements, even on hover */}
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/schemas/Schema.Details.tsx:235]: [question] Is this still applicable?

File: frontend/src/components/pages/connect/helper.tsx [hunk 20]
```diff
@@ -608,14 +546,75 @@ export class ConfirmModal<T> extends Component<{
...
+        okText={$state.error ? 'Retry' : 'Yes'}
+        confirmLoading={$state.isPending}
+        okType="danger"
+        cancelText="No"
+        cancelButtonProps={{ disabled: $state.isPending }}
+        onCancel={cancel}
+        onOk={onOk}
+    >
+        <>
+            <div>
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/connect/helper.tsx:606]: [suggestion] Use Box

File: frontend/src/components/pages/acls/PrincipalGroupEditor.tsx [hunk 21]
```diff
@@ -97,18 +98,21 @@ export const AclPrincipalGroupEditor = observer((p: {
...
-                    );
-                else
-                    message.success(
-                        <>
+                        </span>
+                    });
+                } else {
+                    toast({
+                        status: 'success',
+                        description: <span>
```
Comment (malinskibeniamin) [inline frontend/src/components/pages/acls/PrincipalGroupEditor.tsx:111]: [suggestion] Consider `<Text as="span" ... />`

File: frontend/src/utils/tsxUtils.tsx (hunk 17 above)
Comment (jvorcak) [reply frontend/src/utils/tsxUtils.tsx:567]: Screenshot 2023-09-25 at 14 11 01

I think what is problematic is `GPUQueue API: writeTexture`, we're using `Clipboard API: writeText`

File: frontend/src/components/misc/ErrorBoundary.tsx (hunk 16 above)
Comment (jvorcak) [reply frontend/src/components/misc/ErrorBoundary.tsx:230]: It will as explained here https://github.com/redpanda-data/console/pull/846#discussion_r1335770588

File: frontend/src/components/pages/schemas/Schema.Details.tsx (hunk 19 above)
Comment (jvorcak) [reply frontend/src/components/pages/schemas/Schema.Details.tsx:235]: You mean checking the availability of a clipboard? The comment about `isDisabled` is still applicable.
While I think it's not absolutely necessary to check for an availability of this API, I didn't want to remove functionality as part of this PR. btw it's supported in almost all browsers according to caniuse.

File: frontend/src/components/pages/connect/helper.tsx (hunk 20 above)
Comment (jvorcak) [reply frontend/src/components/pages/connect/helper.tsx:606]: This is not related to a Toast, I'm sure it's addressed in #847

**Review by malinskibeniamin — APPROVED** [review]

**Review by rikimaru0345 — APPROVED** [review]

File: frontend/src/components/misc/HideStatisticsBarButton.tsx (hunk 3 above)
Comment (jvorcak) [reply frontend/src/components/misc/HideStatisticsBarButton.tsx:20]: I think it's more related to the Tooltip? I'm having a hard time to fully-reproduce this specific button to be honest. Maybe I can create a task for it later?

File: frontend/src/components/pages/schemas/Schema.Details.tsx (hunk 12 above)
Comment (rikimaru0345) [reply frontend/src/components/pages/schemas/Schema.Details.tsx:283]: Both are fine, personally I'd use `my` if available.
This is not something worth spending time on / discussing. 
Both do the job, and changing between them is trivial
